feat(packages): add QuestDB driver

puzpuzpuz opened the pull request:
**Check List**
- [x] Tests has been run in packages where changes made if available
- [x] Linter has been run for changed code
- [x] Tests for the changes have been added if not covered yet
- [x] Docs have been added / updated if required

**Description of Changes Made (if issue reference is not provided)**

Adds [QuestDB](https://questdb.io/) driver, as well as some basic documentation for the driver.

Known limitations:
* The driver doesn't support stream API due to a `pg-cursor` + `pg-query-stream` compatibility issue. This should be fixed in one of the next QuestDB releases.

I did some smoke testing based on Cube.js Playground, but I'm open to any suggestions on further driver testing.

cristipp commented on packages/cubejs-questdb-driver/src/QuestQuery.ts:
```diff
@@ -0,0 +1,144 @@
+
+  public unixTimestampSql(): string {
+    // QuestDB's now() function returns epoch timestamp with microsecond granularity.
+
```
Please remove empty line.

puzpuzpuz replied:
Addressed in df4d6df2c84c483d0a6aa5d8c51a306b23d6e9ff

cristipp commented on packages/cubejs-questdb-driver/src/QuestQuery.ts:
```diff
@@ -0,0 +1,144 @@
+  }
+
+  public orderHashToString(hash: any): string | null {
+    // QuestDB has partial support for order by index column, so map these to the alias names.
```
The comment is a bit terse. Perhaps a small example could help?

puzpuzpuz replied:
Added an example in df4d6df2c84c483d0a6aa5d8c51a306b23d6e9ff

cristipp replied:
Thanks! This makes it clear.

cristipp commented on packages/cubejs-questdb-driver/src/QuestQuery.ts:
```diff
@@ -0,0 +1,144 @@
+  }
+
+  public groupByClause(): string {
+    // QuestDB doesn't support group by index column, so map these to the alias names.
```
Can you please provide a brief example of what this is a workaround for? What kind of cube config / query combo leads to a group by index column being issued?

puzpuzpuz replied:
Added an example in df4d6df2c84c483d0a6aa5d8c51a306b23d6e9ff

cristipp replied:
Thanks!

cristipp reviewed (commented):
Initial pass done, looking great!

puzpuzpuz commented on packages/cubejs-questdb-driver/src/QuestDriver.ts:
```diff
@@ -0,0 +1,225 @@
+
+const NativeTypeToQuestType: Record<string, string> = {};
+
+Object.entries(types.builtins).forEach(([key, value]) => {
```
Addressed in df4d6df2c84c483d0a6aa5d8c51a306b23d6e9ff

puzpuzpuz commented on packages/cubejs-questdb-driver/src/QuestDriver.ts:
```diff
@@ -0,0 +1,225 @@
+
+const timestampDataTypes = [
+  // @link TypeId.DATE
+  1082,
```
Addressed in df4d6df2c84c483d0a6aa5d8c51a306b23d6e9ff

`TypeId.DATE` won't work, so I used `types.builtins` instead.

puzpuzpuz commented on packages/cubejs-questdb-driver/src/QuestDriver.ts:
```diff
@@ -0,0 +1,225 @@
+    };
+  }
+
+  protected getInitialConfiguration(): Partial<QuestDriverConfiguration> {
```
Addressed in df4d6df2c84c483d0a6aa5d8c51a306b23d6e9ff

puzpuzpuz commented on packages/cubejs-questdb-driver/src/QuestDriver.ts:
```diff
@@ -0,0 +1,225 @@
+    }
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
```
Addressed in df4d6df2c84c483d0a6aa5d8c51a306b23d6e9ff

puzpuzpuz commented on packages/cubejs-questdb-driver/src/QuestDriver.ts:
```diff
@@ -0,0 +1,225 @@
+    return result.rows;
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
```
Addressed in df4d6df2c84c483d0a6aa5d8c51a306b23d6e9ff

puzpuzpuz commented on packages/cubejs-questdb-driver/src/QuestDriver.ts:
```diff
@@ -0,0 +1,225 @@
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  public async createSchemaIfNotExists(schemaName: string): Promise<any> {
```
Addressed in df4d6df2c84c483d0a6aa5d8c51a306b23d6e9ff

puzpuzpuz commented on packages/cubejs-questdb-driver/src/QuestDriver.ts:
```diff
@@ -0,0 +1,225 @@
+    return metadata;
+  }
+
+  // eslint-disable-next-line camelcase, @typescript-eslint/no-unused-vars
```
Removed unused vars part in df4d6df2c84c483d0a6aa5d8c51a306b23d6e9ff

As for camelcase names, that's a part of the Cube.js driver API. See `DriverInterface.getTablesQuery`

puzpuzpuz commented:
Note: since I tried to reuse the code of other drivers as much as possible in order to keep the code base consistent, many of the review comments are applicable to the existing drivers. Yet, I only changed the QuestDB driver when addressing review comments.

puzpuzpuz commented on packages/cubejs-questdb-driver/src/QuestDriver.ts:
```diff
@@ -0,0 +1,225 @@
+  public async tablesSchema() {
+    const tables = await this.getTablesQuery('');
+
+    const metadata: Record<string, Record<string, object>> = { '': {} };
```
The thing is that QuestDB doesn't have a notion of schema/logical database while this driver method has to return a `{ 'schema_name': { 'table1': {...} } }` object. I tried to use a non-empty hardcoded schema name here (say, 'global'), but it leads to `global.table_name` being used in the generated queries. So, using an empty string here is a mere hack to work around this problem.

If there is already a way to solve it in a proper way, I'm all for it. Otherwise, I could maybe add a `supportsSchema(): boolean` driver method and try to use it in cube-cli and/or dev server.

WDYT?

cristipp replied:
I see. Does '' work, or do you get `.table_name`? If you could add a comment in the source explaining what you already explained above, that would suffice.

puzpuzpuz replied:
Makes sense. Added a comment in 6ca4b88d350e5de897eb4c9aac23fec896fcd923

cristipp commented on packages/cubejs-questdb-driver/test/QuestQuery.test.ts:
```diff
@@ -0,0 +1,172 @@
+    });
+    `, []);
+
+  it('test equal filters', async () => {
```
We don't need this test here, just the ORDER BY / GROUP BY.

puzpuzpuz replied:
I'd prefer to keep this test since it validates that `= NULL` filter expression is being used instead of `IS NULL`. So, it's as valuable as the subsequent ORDER BY / GROUP BY tests.

cristipp commented on packages/cubejs-questdb-driver/test/QuestQuery.test.ts:
```diff
@@ -0,0 +1,172 @@
+      }
+    })
+
+    cube(\`Deals\`, {
```
This cube is unused by the tests, please remove.

puzpuzpuz replied:
Removed in 06b539bf13bffe597f855018d8d80fe383812df3

cristipp commented on packages/cubejs-questdb-driver/test/QuestQuery.test.ts:
```diff
@@ -0,0 +1,172 @@
+      }
+    })
+    
+    cube(\`SalesManagers\`, {
```
This cube is unused by the tests, please remove.

puzpuzpuz replied:
Removed in 06b539bf13bffe597f855018d8d80fe383812df3
